File: js/afterglowpy_api.js
var afterglowpyOK = false;

function afterglowpyReady() {
    console.log("afterglowpy OK to go!");
    afterglowpyOK = true;
}

class AfterglowpyCaller {
    constructor() {
        let sig = [];
        for(let i = 0; i<35; i++) {
            sig.push('number');
        }

        this.c_calcFluxDensity = Module.cwrap('calcFluxDensity',
                                                'number', sig);
        this.c_getVersion = Module.cwrap('getVersion',
                                                'number', []);
        this.c_getGitVersion = Module.cwrap('getGitVersion',
                                                'number', []);
        this.c_getJetTypeCode = Module.cwrap('getJetTypeCode',
                                                'number', ['number']);
        this.c_getEnvTypeCode = Module.cwrap('getEnvTypeCode',
                                                'number', ['number']);
        this.c_getGammaTypeCode = Module.cwrap('getGammaTypeCode',
                                                'number', ['number']);
    }

    getVersion() {

        let ver_buf = this.c_getVersion();
        let ver = UTF8ToString(ver_buf);
        Module._free(ver_buf);

        return ver;
    }

    getGitVersion() {

        let ver_buf = this.c_getGitVersion();
        let ver = UTF8ToString(ver_buf);
        Module._free(ver_buf);

        return ver;
    }

    getJetTypeCode(jetTypeName) {
        let bufSize = 4*jetTypeName.length + 1;
        let c_jetTypeName = Module._malloc(bufSize);
        stringToUTF8(jetTypeName, c_jetTypeName, bufSize);
        
        let jetType = this.c_getJetTypeCode(c_jetTypeName);

        Module._free(c_jetTypeName);

        return jetType;
    }

    getEnvTypeCode(envTypeName) {
        let bufSize = 4*envTypeName.length + 1;
        let c_envTypeName = Module._malloc(bufSize);
        stringToUTF8(envTypeName, c_envTypeName, bufSize);
        
        let envType = this.c_getEnvTypeCode(c_envTypeName);

        Module._free(c_envTypeName);

        return envType;
    }

    getGammaTypeCode(gammaTypeName) {
        let bufSize = 4*gammaTypeName.length + 1;
        let c_gammaTypeName = Module._malloc(bufSize);
        stringToUTF8(gammaTypeName, c_gammaTypeName, bufSize);
        
        let gammaType = this.c_getGammaTypeCode(c_gammaTypeName);

        Module._free(c_gammaTypeName);

        return gammaType;
    }

    fluxDensity(t, nu, Z) {
        let size = 8;
        let N = t.length;

        let c_t = Module._malloc(N * size);
        let c_nu = Module._malloc(N * size);
        let c_Fnu = Module._malloc(N * size);
        for(let i = 0; i<N; i++) {
            Module.setValue(c_t + i*size, t[i] / (1 + Z.z), 'double');
            Module.setValue(c_nu + i*size, nu[i] * (1 + Z.z), 'double');
        }

        console.log("in GRB fluxDensity");
        console.log(Z);

        let err = this.c_calcFluxDensity(c_t, c_nu, c_Fnu, N,
                                Z.dL, Z.thetaV, Z.E0, Z.thetaC, Z.thetaW, Z.b,
                                Z.n0, Z.p, Z.epse, Z.epsB, Z.xiN, -1,
                                Z.envType, 1.0e18, 0, 0, 
                                Z.E0, Z.thetaC, 
                                1000, 5, 7,
                                0.01, 0.01, 0.01, 1000, 1000,
                                Z.jetType, Z.specType,
                                7, 0, 0);

        if(err > 0) {
            console.log("There was an error!");
        }

        let Fnu = [];

        for(let i = 0; i<N; i++) {
            Fnu.push((1+Z.z) * Module.getValue(c_Fnu + i*size, 'double'));
        }

        Module._free(c_t);
        Module._free(c_nu);
        Module._free(c_Fnu);

    return Fnu
    }
}

File: src/afterglowpy_api.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <emscripten.h>
#include "offaxis_struct.h"
#include "shockEvolution.h"

char * getVersion()
{
    int N = strlen(AFTERGLOWPY_VERSION);
    char *str = (char *) malloc((N+1) * sizeof(char));
    strncpy(str, AFTERGLOWPY_VERSION, N);
    str[N] = '\0';
    return str;
}

char * getGitVersion()
{
    int N = strlen(GIT_VERSION);
    char *str = (char *) malloc((N+1) * sizeof(char));
    strncpy(str, GIT_VERSION, N);
    str[N] = '\0';
    return str;
}

int getJetTypeCode(char *jetTypeStr)
{
    printf("%s\n", jetTypeStr);
    if(strcmp(jetTypeStr, "Cone") == 0)
        return _cone;
    else if(strcmp(jetTypeStr, "Top Hat") == 0)
        return _tophat;
    else if(strcmp(jetTypeStr, "Gaussian") == 0)
        return _Gaussian;
    else if(strcmp(jetTypeStr, "Power Law Core") == 0)
        return _powerlaw_core;
    else if(strcmp(jetTypeStr, "Gaussian Core") == 0)
        return _Gaussian_core;
    else if(strcmp(jetTypeStr, "Power Law") == 0)
        return _powerlaw;

    return -3;
}

int getEnvTypeCode(const char envTypeStr[])
{
    if(strcmp(envTypeStr, "ISM") == 0)
        return ENV_ISM;
    else if(strcmp(envTypeStr, "Wind") == 0)
        return ENV_WIND;
    else if(strcmp(envTypeStr, "Power Law") == 0)
        return ENV_PL;

    return -1;
}

int getGammaTypeCode(const char gammaTypeStr[])
{
    if(strcmp(gammaTypeStr, "Inf") == 0)
        return GAMMA_INF;
    else if(strcmp(gammaTypeStr, "Flat") == 0)
        return GAMMA_FLAT;
    else if(strcmp(gammaTypeStr, "Even Mass") == 0)
        return GAMMA_EVENMASS;

    return -1;
}

int calcFluxDensity(double *t, double *nu, double *Fnu, int N,
                    double d_L, double thetaV, double E0,
                    double thetaC, double thetaW, double b,
                    double n0, double p, double epse, double epsB,
                    double xiN, double g0,
                    int envType, double R0_env, double k_env, double rho1_env,
                    double E0_glob, double thetaC_glob,
                    int tRes, int latRes, int int_type,
                    double rtol_struct, double rtol_phi, double rtol_theta,
                    int nmax_phi, int nmax_theta,
                    int jet_type, int spec_type,
                    int spread, int counterjet, int gamma_type)
{

    double ta = t[0];
    double tb = t[0];

    int i;
    for(i=1; i<N; i++)
    {
        if(t[i] < ta)
            ta = t[i];
        if(t[i] > tb)
            tb = t[i];
    }

    /*
    printf("%d %d\n", jet_type, spec_type);
    printf("%.3f %.3e %.3f %.3e\n", thetaV, E0, thetaC, d_L);
    printf("%.3e %.3f %.3e %.3e %.3e\n", n0, p, epse, epsB, xiN);
    printf("%d %d %d\n", tRes, latRes, int_type);
    printf("%.3e %.3e %.3e\n", rtol_struct, rtol_phi, rtol_theta);
    printf("%d %d\n", nmax_phi, nmax_theta);
    printf("%d %d %d\n", spread, counterjet, gamma_type);
    */

    struct fluxParams fp;
    setup_fluxParams(&fp, d_L, thetaV, E0, thetaC, thetaW, b,
                        0, 0, 0, 0,
                        n0, p, epse, epsB, xiN, g0, 
                        envType, R0_env, k_env, rho1_env,
                        E0_glob, thetaC_glob, ta, tb,
                        tRes, latRes, int_type,
                        rtol_struct, rtol_phi, rtol_theta,
                        nmax_phi, nmax_theta,
                        spec_type, NULL, 0,
                        spread, counterjet, gamma_type);

    calc_flux_density(jet_type, spec_type, t, nu, Fnu, NULL, N, &fp);

    if(fp.error)
    {
        printf("ERROR\n");
        printf("%s\n", fp.error_msg);
        free_fluxParams(&fp);
        return 1;
    }

    free_fluxParams(&fp);

    return 0;
}

int main(int argc, char *argv[])
{
    EM_ASM(afterglowpyReady());
}
